pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -172,60 +172,95 @@ func (m *ManagerImpl) GetNUMAHints(resource string, amount int) numamanager.Numa
+                		}					
+			}						
+		}
+	}	
+	glog.Infof("[device-manager] Sockets with device: %v", socketValues)
+	
+	var count, devicesTotal int64 = 0, 0 	
+	var amount64 int64
+	amount64 = int64(amount)	
+	var availableSockets []int64
+	var notEmptySkts []int64
+	// Loop through device counts (duplicates of each socket) - return if resources not available on a socket
+	for socket, devicesPerSkt := range duplicate_frequency{
+ 		glog.Infof("[device-manager] Socket : %d , number of devices : %d", socket, devicesPerSkt)
+		devicesTotal = devicesTotal + devicesPerSkt
+		if devicesPerSkt !=0 {
+			notEmptySkts = append(notEmptySkts, socket)
+		}
+		if devicesPerSkt >= amount64 {
+			availableSockets = append(availableSockets, socket)
+			count++
+		}		
+ 	}
+	glog.Infof("[device-manager] Device Count Total: %v", devicesTotal)	
+
+	if count == 0 {
+		glog.Infof("[device-manager] No Socket has available resources - Preferred only")
+	}    
+	glog.Infof("[device-manager] Sockets with 1 or more available devices: %v", notEmptySkts)
+	glog.Infof("[device-manager] Sockets with enough available devices: %v", availableSockets)
+	
+	// Find the largest socket that is NOT empty
+        var largestSkt int64 = 0
+        for i := 0; i < len(notEmptySkts); i++ {
+                if largestSkt < notEmptySkts[i] {
+                        largestSkt = notEmptySkts[i]
+                }
+        }
+        glog.Infof("[device-manager] Largest socket that is NOT empty: %v", largestSkt)
+	
+	var socketMask []int64
+	fullMask:= make([][]int64,len(availableSockets))
+	var socket int64 = 0
+	var i int64
+	// Use largest socket to construct 2D slice (mask) for return to NUMA Manager
+	for j := 0; j < len(availableSockets); j++ {
+		socketMask = nil
+		socket = availableSockets[j]	
+		for i = 0; i <= largestSkt; i++ {
+            		if i == socket {
+                		socketMask = append(socketMask, 1)
+            		} else {
+                		socketMask = append(socketMask, 0)
             		}
-        	}
-    	}
-    	//WIP for preferred  
-    	deviceMask := ""
-    	for _, bit := range defaultMask {
-        	deviceMask += strconv.Itoa(bit)
-    	}
-    	glog.Infof("DeviceMask: %v", deviceMask)
-    	deviceMaskInt, err := strconv.Atoi(deviceMask)
-    	if err != nil {
-        	glog.Errorf("Err: Cannot convert deviceMask to int. %v")
-    	}
-    	nm = append(nm, int64(deviceMaskInt))
-    	glog.Infof("[devicemanager] NUMA Affinities for %v %v resource(s) are %v", amount, resource, nm)
-	return numamanager.NumaMask{
-		  Mask:           nm,
-		  Affinity:       true,
+        	}			
+		fullMask[j] = socketMask 
+	}	
+	
+	//Build final slice for full mask - overall affinity for preferred state (only if more than 1 available skt)
+	overallMask := make([]int64, largestSkt+1)
+	if (len(fullMask) > 1) || ((len(fullMask) == 0) && (devicesTotal >= amount64)) {
+		var k int64
+		for k = 0; k < largestSkt+1; k ++ {
+			for l := 0; l < len(notEmptySkts); l ++ {
```

Comment:
Can use that array here as well.
for _, nonEmptySocket := range notEmptySockets {
    if k == nonEmptySocket
}